add docs for testing with imagetest (#2196)
chainguard-images/images: TERRAFORM.md

These tests may install a Helm chart and/or issue `kubectl` commands against the cluster.
The tests use the config file at `$HOME/.kube/config` to connect to the cluster.

There are two classes of tests that exist today, `oci_exec_test` and `imagetest`.

#### Testing with `oci_exec_test`

> Tests using this resource execute scripts on the host machine. They are extremely simple to author. A minimal example is below:

Reviewer: ```suggestion
Tests using this resource execute scripts on the host machine. They are extremely simple to author. The currently built image is available to the script via the `IMAGE_NAME` environment variable.

A minimal example is below:
```

```hcl
data "oci_exec_test" "foo" {
  # A required image digest. This will be injected as environment variables to
  # whatever script is executed.

>   digest = "my-image@"

Reviewer: +1

  # A script that executes on the host with various context achieved and

>   # plubmed through the provided digest as environment variables.

Reviewer: ```suggestion
  # plumbed through the provided digest as environment variables.
```

  script = <<EOF
    ...
  EOF
}
```

Some examples of these tests can be seen in

* [`ko`](./images/ko/tests/main.tf)
* [check reproducibility](./tflib/publisher/main.tf)

#### Testing with `imagetest`

`imagetest` is a custom provider built explicitly for the `images` release
process. The goal of `imagetest` is to simplify test authoring for both basic
and complex testing scenarios, while providing a safer, more abstract test
execution.

Part of the design of `imagetest` is that all resources created are ephemeral
by default, meaning unlike `oci_exec_test` which executes on the host machine
and requires explicit opt-in resource cleanup, everything created by
`imagetest` will be removed at completion. This makes test authoring simpler,
as any side effects or teardown logic is mostly irrelevant.

Since harnesses typically involve certain host tooling/configuration, please
skim the [local setup](#local-setup) section for the one-time configuration required for local
development.

A minimal test implemented with `imagetest` is below, commented where applicable:

```hcl

> # A _required_ data source, used for keeping a provider internal inventory of #
> # which harnesses should be cleaned up and when. Its important that this should

Reviewer: ```suggestion
# A _required_ data source, used for keeping a provider internal inventory of
# which harnesses should be cleaned up and when. It's important to note that this should
```

# always be fully determined at _plan_ time. In other words, it should have no
# _apply_ time dependencies. To make this hard to mess up, this data source takes
# no inputs. The name of this data source does not matter, and multiple data
# sources can exist within the same terraform scope.
data "imagetest_inventory" "this" {}

# The imagetest provider provides multiple imagetest_harness_* resources that
# abstract the creation and configuration of various test harnesses. These test
# harnesses will be ephemeral to the lifecycle of the _apply_, and are typically

> # authored with sane defaults that require minimal configuration.

Reviewer: ```suggestion
# authored with sensible defaults that require minimal configuration.
```

  # provided inventory. For example, defining another named "foo" harness with
  # the same `data.imagetest_inventory.this` is invalid, but another "foo"
  # harness with `data.imagetest_inventory.that` is valid and often normal,
  # such as when working with for_each'd modules.
  name = "foo"
}

# All tests run within a `imagetest_feature` resource, and are a set of
# instructions that run to completion. The way they are run and how they are run
# vary depending on the provided harness.
resource "imagetest_feature" "foo" {
  # A feature must have a harness, and much like the required inventory field
  # in the harness, this field is structured in a way to take the full data
  # structure of the harness (and not the ID) to make it harder to mess up. The
  # harness passed to the feature dictates _how_ the steps in the feature are
  # executed. These are documented in each of the available imagetest_harness_*
  # resources
  harness = imagetest_harness_k3s.foo

  # The name of the feature. This must be unique within a given harness.
  name = "Footure"
  # The optional description of the harness
  description = "My great footure"

>   # Steps are a series of commands executed in order to completion. Depending #
>   # on the provided harness, these location and method of how these steps are #

Reviewer: ```suggestion
  # Steps are a series of commands executed in order to completion. Depending
  # on the provided harness, these location and method of how these steps are
```

  # executed will vary, but in general, they are typically abstracted to operate

>   # # as shell scripts. For example, in this example, the `imagetest_harness_k3s`

Reviewer: ```suggestion
  # as shell scripts. For example, in this example, the `imagetest_harness_k3s`
```

  # is used, which means these steps will execute within a sandboxed `wolfi-base`
  # image that has `kubectl` pre-installed and pre-configured to connect to the
  # network attached k3s cluster. The intent with the harness feature setups is
  # always to absolve as boilerplate setup as possible from the test authors.
  steps = [
    {
      name = "Do a thing"
      cmd = <<EOF
        kubectl get po -A
      EOF
    },
    # Where applicable, additional step fields are available to hoist commonly
    # performed actions out of shell scripts and into the provider. The example
    # below shows an optional retry = {} block, which abstracts the common
    # approach of a bash retry loop with a simple provider resource definition
    {
      name = "Do a thing that may flake"
      cmd = <<EOF
        ...
      EOF

>       # Retry with a linear backoff
>       retry = { attempts = 5, delay = "10s" },
>       # Retry with an exponential backoff
>       # retry = { attempts = 3, delay = "5s", factor = "2" },

Reviewer: This can be a little confusing to read when a separate example is commented out:
```suggestion
      # Retry with a linear backoff
      retry = { attempts = 5, delay = "10s" },
      # Retry with an exponential backoff
      retry = { attempts = 3, delay = "5s", factor = "2" },
```

Perhaps break it down into two examples?
